kong/reports.lua

```diff
@@ -150,6 +156,83 @@ local function create_timer(...)
 end
 
 
+local function get_counter(key)
+  local count, err = kong_dict:get(key)
+  if err then
+    log(WARN, "could not get " .. key .. " from 'kong' shm: ", err)
+  end
+  return count or 0
+end
+
+
+local function reset_counter(key, amount)
+  local ok, err = kong_dict:incr(key, -amount, amount)
```

Comment:
Better use `:safe_set(key, 0)` to avoid race-conditions in between the previous `get()` and this `incr()` step.

Reply:
It puzzled me a bit too, but I think we *want* to preserve these "race conditions".

This is my understanding of what's happening now:

```
                Flow of Time
                    ||| 
                    VVV 

      Worker A       |     Worker B
                     |
                     |
get_counter -> 100   |
                     |  <log phase> incr_counter(1) -> 101
                     |
reset_counter(-100)  |
============================================
                     
Final counter value: 1 (correct, the incr_counter from Worker B was preserved)
```

Now with what you propose:
```
                Flow of Time
                    ||| 
                    VVV 

      Worker A       |     Worker B
                     |
                     |
get_counter -> 100   |
                     |  <log phase> incr_counter(1) -> 101
                     |
safe_set(0)          |
============================================
                     
Final counter value: 0 (incorrect, the incr_counter from Worker B was lost)
```

Reply:
@kikito You are correct! My bad. I should know, this was my implementation -_- https://github.com/Kong/kong/commit/2ace69179fba2b3c193bd3cfb73f4102514ab306#diff-10418ffd13cdf2bc868b10cce41a8781R169

Reply:
I have added the explanation as a comment in the code. Hopefully it will help whomever visits this code in another 3 years :)